File: src/core/contract-ai/evolution/state-analyzer.ts
/**
 * State Analyzer
 * 
 * Multi-level state analysis: Contract ↔ SourceCode ↔ Service ↔ Logs
 * 
 * @version 2.4.1
 */

import * as fs from 'fs';
import * as path from 'path';
import { ContractAI } from '../types';

// ============================================================================
// TYPES
// ============================================================================

export interface StateDiscrepancy {
  level: 'contract-code' | 'code-service' | 'service-logs';
  severity: 'info' | 'warning' | 'error';
  source: string;
  expected: string;
  actual: string;
  suggestion: string;
}

export interface MultiLevelState {
  contract: {
    entities: string[];
    endpoints: string[];
    targets: string[];
  };
  sourceCode: {
    files: string[];
    detectedEndpoints: string[];
    detectedEntities: string[];
  };
  service: {
    port: number;
    running: boolean;
    healthEndpoint: boolean;
    respondingEndpoints: string[];
    probes: Array<{ endpoint: string; ok: boolean; status?: number; error?: string }>;
  };
  logs: {
    errors: string[];
    warnings: string[];
    lastActivity: string | null;
  };
  discrepancies: StateDiscrepancy[];
  reconciled: boolean;
}

// ============================================================================
// STATE ANALYZER CLASS
// ============================================================================

export class StateAnalyzer {
  private outputDir: string;
  private port: number;

  constructor(outputDir: string, port: number) {
    this.outputDir = outputDir;
    this.port = port;
  }

  setPort(port: number): void {
    this.port = port;
  }

  analyzeContract(): { entities: string[]; endpoints: string[]; targets: string[] } {
    const contractPath = path.join(this.outputDir, 'contract', 'contract.ai.json');
    if (!fs.existsSync(contractPath)) {
      return { entities: [], endpoints: [], targets: [] };
    }

    try {
      const contract = JSON.parse(fs.readFileSync(contractPath, 'utf-8')) as ContractAI;
      const entities = (contract.definition?.entities || []).map(e => e.name);
      const targets = Array.from(new Set(
        (contract.generation?.instructions || [])
          .map(i => i.target)
          .filter(t => t && t !== 'all')
      ));
      
      const apiPrefix = contract.definition?.api?.prefix || '/api/v1';
      const resources: any[] = (contract.definition?.api as any)?.resources || [];

      const endpoints: string[] = ['GET /health'];

      for (const r of resources) {
        const name = String(r?.name || '').trim();
        if (!name) continue;
        const base = `${apiPrefix}/${name}`.replace(/\/+/g, '/');
        const ops: string[] = Array.isArray(r?.operations) ? r.operations : [];

        if (ops.includes('list')) endpoints.push(`GET ${base}`);
        if (ops.includes('create')) endpoints.push(`POST ${base}`);
        if (ops.includes('get')) endpoints.push(`GET ${base}/:id`);
        if (ops.includes('update')) endpoints.push(`PUT ${base}/:id`);
        if (ops.includes('delete')) endpoints.push(`DELETE ${base}/:id`);

        const custom: any[] = Array.isArray(r?.customEndpoints) ? r.customEndpoints : [];
        for (const c of custom) {
          const method = String(c?.method || 'GET').toUpperCase();
          const rel = String(c?.path || '').startsWith('/') ? String(c?.path || '') : `/${String(c?.path || '')}`;
          const full = `${base}${rel}`.replace(/\/+/g, '/');
          endpoints.push(`${method} ${full}`);
        }
      }

      return { entities, endpoints, targets };
    } catch {
      return { entities: [], endpoints: [], targets: [] };
    }
  }

  analyzeSourceCode(): { files: string[]; detectedEndpoints: string[]; detectedEntities: string[] } {
    const files: string[] = [];
    const detectedEndpoints: string[] = [];
    const detectedEntities: string[] = [];

    const apiDir = path.join(this.outputDir, 'api');
    if (fs.existsSync(apiDir)) {
      this.scanDir(apiDir, files);
    }

    const serverPath = path.join(this.outputDir, 'api', 'src', 'server.ts');
    if (fs.existsSync(serverPath)) {
      const content = fs.readFileSync(serverPath, 'utf-8');
      
      const routePatterns = content.matchAll(/app\.(get|post|put|delete|patch)\s*\(\s*['"`]([^'"`]+)['"`]/gi);
      for (const match of routePatterns) {
        detectedEndpoints.push(`${match[1].toUpperCase()} ${match[2]}`);
      }

      const entityPatterns = content.matchAll(/(?:const|let)\s+(\w+)s?\s*(?:=\s*new\s+Map|:\s*\w+\[\]|=\s*\[\])/gi);
      for (const match of entityPatterns) {
        const name = match[1].charAt(0).toUpperCase() + match[1].slice(1);
        if (!['Next', 'Result', 'Error'].includes(name)) {
          detectedEntities.push(name);
        }
      }

      const interfacePatterns = content.matchAll(/interface\s+(\w+)\s*\{/g);
      for (const match of interfacePatterns) {
        if (!['Request', 'Response', 'TestResult'].includes(match[1])) {
          detectedEntities.push(match[1]);
        }
      }
    }

    return { 
      files: files.map(f => f.replace(this.outputDir + '/', '')), 
      detectedEndpoints: [...new Set(detectedEndpoints)],
      detectedEntities: [...new Set(detectedEntities)]
    };
  }

  private scanDir(dir: string, files: string[]): void {
    try {
      for (const entry of fs.readdirSync(dir)) {
        const fullPath = path.join(dir, entry);
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory() && !entry.includes('node_modules')) {
          this.scanDir(fullPath, files);
        } else if (stat.isFile()) {
          files.push(fullPath);
        }
      }
    } catch {
      // Ignore errors
    }
  }

  async analyzeService(endpointsToProbe: string[]): Promise<{ running: boolean; healthEndpoint: boolean; respondingEndpoints: string[]; probes: Array<{ endpoint: string; ok: boolean; status?: number; error?: string }>; port: number }> {
    const result = { running: false, healthEndpoint: false, respondingEndpoints: [] as string[], probes: [] as Array<{ endpoint: string; ok: boolean; status?: number; error?: string }>, port: this.port };

    const probe = async (endpoint: string, method: string, urlPath: string, timeoutMs: number): Promise<void> => {
      try {
        const res = await fetch(`http://localhost:${this.port}${urlPath}`, {
          method,
          signal: AbortSignal.timeout(timeoutMs)
        });
        const ok = res.ok;
        result.probes.push({ endpoint, ok, status: res.status });
        if (ok) result.respondingEndpoints.push(endpoint);
      } catch (e: any) {
        result.probes.push({ endpoint, ok: false, error: (e && e.message) ? String(e.message) : String(e) });
      }
    };

    // Health probe defines "running" (server responded)
    try {
      const res = await fetch(`http://localhost:${this.port}/health`, { signal: AbortSignal.timeout(2000) });
      result.running = true;
      result.healthEndpoint = res.ok;
      result.probes.push({ endpoint: 'GET /health', ok: res.ok, status: res.status });
      if (res.ok) result.respondingEndpoints.push('GET /health');
    } catch (e: any) {
      result.probes.push({ endpoint: 'GET /health', ok: false, error: (e && e.message) ? String(e.message) : String(e) });
      return result;
    }

    const unique = Array.from(new Set(endpointsToProbe || []));
    const candidates = unique
      .map(s => String(s || '').trim())
      .filter(s => s.includes(' '))
      .slice(0, 20);

    for (const ep of candidates) {
      const [methodRaw, pathRaw] = ep.split(' ', 2);
      const method = String(methodRaw || '').toUpperCase();
      const urlPath = String(pathRaw || '').trim();
      if (method !== 'GET') continue;
      if (!urlPath.startsWith('/')) continue;
      if (urlPath.includes(':')) continue;
      if (urlPath === '/health') continue;
      await probe(`GET ${urlPath}`, 'GET', urlPath, 1500);
    }

    return result;
  }

  analyzeLogs(): { errors: string[]; warnings: string[]; lastActivity: string | null } {
    const logsDir = path.join(this.outputDir, 'logs');
    const errors: string[] = [];
    const warnings: string[] = [];
    let lastActivity: string | null = null;

    if (!fs.existsSync(logsDir)) {
      return { errors, warnings, lastActivity };
    }

    try {
      const logFiles = fs.readdirSync(logsDir)
        .filter(f => f.endsWith('.rcl.md') || f.endsWith('.log'))
        .map(f => ({
          file: f,
          mtime: fs.statSync(path.join(logsDir, f)).mtimeMs
        }))
        .sort((a, b) => a.mtime - b.mtime)
        .slice(-2)
        .map(x => x.file);

      for (const file of logFiles) {
        const content = fs.readFileSync(path.join(logsDir, file), 'utf-8');
        const tail = content.split('\n').slice(-250);

        for (const line of tail) {
          const t = (line || '').trim();
          if (!t) continue;
          if (t.startsWith('```')) continue;
          if (t.startsWith('# @type')) continue;
          if (t.includes('[error]') || t.startsWith('Error:') || t.includes('EADDRINUSE') || t.includes('TSError') || t.includes('Exception')) {
            errors.push(t.substring(0, 160));
          }
          if (t.includes('[warn]') || t.startsWith('Warning:') || t.startsWith('WARN') || t.includes('Deprecation')) {
            warnings.push(t.substring(0, 160));
          }

          const m = t.match(/(\d{4}-\d{2}-\d{2}T[\d:]+)/);
          if (m) lastActivity = m[1];
        }
      }
    } catch {
      // Ignore errors
    }

    return { errors: errors.slice(-10), warnings: warnings.slice(-10), lastActivity };
  }

  findDiscrepancies(
    contract: ReturnType<typeof this.analyzeContract>,
    sourceCode: ReturnType<typeof this.analyzeSourceCode>,
    service: Awaited<ReturnType<typeof this.analyzeService>>,
    logs: ReturnType<typeof this.analyzeLogs>
  ): StateDiscrepancy[] {
    const discrepancies: StateDiscrepancy[] = [];

    const probeMap = new Map<string, { endpoint: string; ok: boolean; status?: number; error?: string }>();
    for (const p of (service as any).probes || []) {
      probeMap.set(String(p.endpoint), p);
    }

    for (const entity of contract.entities) {
      const found = sourceCode.detectedEntities.some(e => 
        e.toLowerCase() === entity.toLowerCase() || 
        e.toLowerCase().includes(entity.toLowerCase())
      );
      if (!found) {
        discrepancies.push({
          level: 'contract-code',
          severity: 'error',
          source: 'contract.entities',
          expected: `Entity "${entity}" defined in contract`,
          actual: `Entity not found in source code`,
          suggestion: `Add ${entity} model and CRUD endpoints to server.ts`
        });
      }
    }

    for (const endpoint of contract.endpoints) {
      const found = sourceCode.detectedEndpoints.some(e => 
        e.toLowerCase().includes(endpoint.split(' ')[1]?.toLowerCase() || '')
      );
      if (!found && !endpoint.includes(':id')) {
        discrepancies.push({
          level: 'contract-code',
          severity: 'warning',
          source: 'contract.endpoints',
          expected: endpoint,
          actual: 'Endpoint not found in source code',
          suggestion: `Add ${endpoint} handler to server.ts`
        });
      }
    }

    if (service.running) {
      // Health must respond if service is marked running
      if (!service.healthEndpoint) {
        const healthProbe = probeMap.get('GET /health');
        discrepancies.push({
          level: 'code-service',
          severity: 'error',
          source: 'service.health',
          expected: 'GET /health should respond',
          actual: healthProbe?.error ? `health probe error: ${healthProbe.error}` : `health status: ${healthProbe?.status ?? 'unknown'}`,
          suggestion: 'Ensure the server registers /health and is listening on the configured port'
        });
      }

      const expectedGets = contract.endpoints
        .map(e => String(e || '').trim())
        .filter(e => e.startsWith('GET '))
        .filter(e => !e.includes(':id'))
        .filter(e => e !== 'GET /health');

      for (const ep of expectedGets.slice(0, 10)) {
        // Only validate endpoints that exist in source code (avoid false positives from contract)
        const existsInCode = sourceCode.detectedEndpoints.includes(ep);
        if (!existsInCode) continue;

        const probe = probeMap.get(ep);
        if (!probe || !probe.ok) {
          const actual = probe
            ? (probe.error ? `probe error: ${probe.error}` : `status: ${probe.status ?? 'unknown'}`)
            : 'not probed';
          discrepancies.push({
            level: 'code-service',
            severity: 'warning',
            source: 'service.probe',
            expected: `${ep} should respond`,
            actual,
            suggestion: 'Check route registration, API prefix, and that the service is still running'
          });
        }
      }
    } else {
      discrepancies.push({
        level: 'code-service',
        severity: 'error',
        source: 'service.running',
        expected: 'Service should be running',
        actual: 'Service is not running',
        suggestion: 'Start the service with npm start or check for startup errors'
      });
    }

    if (logs.errors.length > 0) {
      const criticalPatterns = [
        'TSError',
        'Unable to compile TypeScript',
        'Cannot find module',
        'Cannot find name',
        'UnhandledPromiseRejection',
        'ReferenceError',
        'TypeError',
        'SyntaxError',
        'Exception'
      ];
      const critical = logs.errors.filter(e => criticalPatterns.some(p => String(e).includes(p)));
      const severity: StateDiscrepancy['severity'] = critical.length > 0 ? 'error' : 'warning';
      const sample = (critical.length > 0 ? critical : logs.errors)[(critical.length > 0 ? critical : logs.errors).length - 1];
      discrepancies.push({
        level: 'service-logs',
        severity,
        source: 'logs.errors',
        expected: severity === 'error' ? 'No critical errors in logs' : 'No recent warnings in logs',
        actual: `${logs.errors.length} log issues detected. Example: ${String(sample || '').substring(0, 160)}`,
        suggestion: severity === 'error'
          ? 'Fix the critical errors shown in logs'
          : 'Investigate warnings; if they are historical, clear logs or rotate to reduce noise'
      });
    }

    return discrepancies;
  }

  async analyze(): Promise<MultiLevelState> {
    const contract = this.analyzeContract();
    const sourceCode = this.analyzeSourceCode();
    const service = await this.analyzeService(sourceCode.detectedEndpoints);
    const logs = this.analyzeLogs();
    const discrepancies = this.findDiscrepancies(contract, sourceCode, service, logs);

    return {
      contract,
      sourceCode,
      service,
      logs,
      discrepancies,
      reconciled: discrepancies.filter(d => d.severity === 'error').length === 0
    };
  }

  generateReconciliationPlan(state: MultiLevelState): string[] {
    const plan: string[] = [];

    const contractCode = state.discrepancies.filter(d => d.level === 'contract-code');
    const codeService = state.discrepancies.filter(d => d.level === 'code-service');
    const serviceLogs = state.discrepancies.filter(d => d.level === 'service-logs');

    if (contractCode.length > 0) {
      plan.push('Contract ↔ Code reconciliation needed:');
      for (const d of contractCode) {
        plan.push(`  - ${d.suggestion}`);
      }
    }

    if (codeService.length > 0) {
      plan.push('Code ↔ Service reconciliation needed:');
      for (const d of codeService) {
        plan.push(`  - ${d.suggestion}`);
      }
    }

    if (serviceLogs.length > 0) {
      plan.push('Service ↔ Logs reconciliation needed:');
      for (const d of serviceLogs) {
        plan.push(`  - ${d.suggestion}`);
      }
    }

    if (plan.length === 0) {
      plan.push('All levels are reconciled - no discrepancies found');
    }

    return plan;
  }
}

export default StateAnalyzer;

File: src/python/reclapp/evolution/state_analyzer.py
"""
State Analyzer Module

Multi-level state analysis for evolution pipeline.
Tracks contract, code, and runtime state discrepancies.

Mirrors: src/core/contract-ai/evolution/state-analyzer.ts
"""

import json
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any, Optional


class StateLevel(str, Enum):
    """State analysis levels"""
    CONTRACT = "contract"
    CODE = "code"
    RUNTIME = "runtime"


class DiscrepancyType(str, Enum):
    """Types of state discrepancies"""
    MISSING_ENTITY = "missing_entity"
    MISSING_FIELD = "missing_field"
    TYPE_MISMATCH = "type_mismatch"
    MISSING_ENDPOINT = "missing_endpoint"
    MISSING_FILE = "missing_file"
    SCHEMA_MISMATCH = "schema_mismatch"
    RUNTIME_ERROR = "runtime_error"


@dataclass
class StateDiscrepancy:
    """A discrepancy between expected and actual state"""
    level: StateLevel
    type: DiscrepancyType
    path: str
    expected: Any
    actual: Any
    message: str
    severity: str = "error"
    
    def to_dict(self) -> dict:
        return {
            "level": self.level.value,
            "type": self.type.value,
            "path": self.path,
            "expected": self.expected,
            "actual": self.actual,
            "message": self.message,
            "severity": self.severity
        }


@dataclass
class ContractState:
    """State extracted from contract"""
    app_name: str = ""
    version: str = "1.0.0"
    entities: list[dict] = field(default_factory=list)
    endpoints: list[dict] = field(default_factory=list)
    tech_stack: dict = field(default_factory=dict)


@dataclass
class CodeState:
    """State extracted from generated code"""
    files: list[str] = field(default_factory=list)
    models: list[str] = field(default_factory=list)
    routes: list[str] = field(default_factory=list)
    has_package_json: bool = False
    has_tsconfig: bool = False


@dataclass
class RuntimeState:
    """State extracted from running service"""
    is_running: bool = False
    port: int = 0
    health_ok: bool = False
    endpoints_responding: list[str] = field(default_factory=list)


@dataclass
class MultiLevelState:
    """Combined multi-level state"""
    contract: ContractState = field(default_factory=ContractState)
    code: CodeState = field(default_factory=CodeState)
    runtime: RuntimeState = field(default_factory=RuntimeState)
    discrepancies: list[StateDiscrepancy] = field(default_factory=list)
    
    def is_consistent(self) -> bool:
        """Check if all levels are consistent"""
        return len([d for d in self.discrepancies if d.severity == "error"]) == 0
    
    def to_dict(self) -> dict:
        return {
            "contract": {
                "app_name": self.contract.app_name,
                "version": self.contract.version,
                "entities": len(self.contract.entities),
                "endpoints": len(self.contract.endpoints)
            },
            "code": {
                "files": len(self.code.files),
                "models": len(self.code.models),
                "routes": len(self.code.routes)
            },
            "runtime": {
                "is_running": self.runtime.is_running,
                "health_ok": self.runtime.health_ok
            },
            "discrepancies": [d.to_dict() for d in self.discrepancies],
            "is_consistent": self.is_consistent()
        }


class StateAnalyzer:
    """
    Analyzes state across contract, code, and runtime levels.
    
    Mirrors TypeScript StateAnalyzer for consistency checking.
    
    Example:
        analyzer = StateAnalyzer(target_dir)
        state = await analyzer.analyze(contract)
        
        if not state.is_consistent():
            for d in state.discrepancies:
                print(f"{d.level}: {d.message}")
    """
    
    def __init__(self, target_dir: str, verbose: bool = False):
        self.target_dir = Path(target_dir)
        self.verbose = verbose
    
    async def analyze(self, contract: Optional[dict] = None) -> MultiLevelState:
        """Perform full multi-level analysis"""
        state = MultiLevelState()
        
        # Analyze contract level
        if contract:
            state.contract = self._analyze_contract(contract)
        else:
            contract = self._load_contract()
            if contract:
                state.contract = self._analyze_contract(contract)
        
        # Analyze code level
        state.code = self._analyze_code()
        
        # Analyze runtime level
        state.runtime = await self._analyze_runtime()
        
        # Find discrepancies
        state.discrepancies = self._find_discrepancies(state)
        
        return state
    
    def _load_contract(self) -> Optional[dict]:
        """Load contract from file"""
        contract_path = self.target_dir / "contract" / "contract.ai.json"
        if contract_path.exists():
            try:
                return json.loads(contract_path.read_text())
            except:
                pass
        return None
    
    def _analyze_contract(self, contract: dict) -> ContractState:
        """Extract state from contract"""
        state = ContractState()
        
        definition = contract.get("definition", {})
        app = definition.get("app", {})
        
        state.app_name = app.get("name", "")
        state.version = app.get("version", "1.0.0")
        state.entities = definition.get("entities", [])
        
        # Extract endpoints from API definition
        api = definition.get("api", {})
        resources = api.get("resources", [])
        for resource in resources:
            if isinstance(resource, dict):
                state.endpoints.extend(resource.get("endpoints", []))
            elif isinstance(resource, str):
                # Generate CRUD endpoints for resource
                resource_lower = resource.lower()
                state.endpoints.extend([
                    {"method": "GET", "path": f"/api/{resource_lower}"},
                    {"method": "POST", "path": f"/api/{resource_lower}"},
                    {"method": "GET", "path": f"/api/{resource_lower}/:id"},
                    {"method": "PUT", "path": f"/api/{resource_lower}/:id"},
                    {"method": "DELETE", "path": f"/api/{resource_lower}/:id"},
                ])
        
        generation = contract.get("generation", {})
        state.tech_stack = generation.get("techStack", {})
        
        return state
    
    def _analyze_code(self) -> CodeState:
        """Extract state from generated code"""
        state = CodeState()
        
        # Find all files
        if self.target_dir.exists():
            for path in self.target_dir.rglob("*"):
                if path.is_file() and "node_modules" not in str(path):
                    rel_path = str(path.relative_to(self.target_dir))
                    state.files.append(rel_path)
                    
                    # Categorize files
                    if "/models/" in rel_path or rel_path.endswith(".model.ts"):
                        state.models.append(rel_path)
                    elif "/routes/" in rel_path or rel_path.endswith(".routes.ts"):
                        state.routes.append(rel_path)
        
        # Check for key files
        state.has_package_json = (self.target_dir / "api" / "package.json").exists() or \
                                  (self.target_dir / "package.json").exists()
        state.has_tsconfig = (self.target_dir / "api" / "tsconfig.json").exists() or \
                              (self.target_dir / "tsconfig.json").exists()
        
        return state
    
    async def _analyze_runtime(self) -> RuntimeState:
        """Extract state from running service"""
        import urllib.request
        import urllib.error
        
        state = RuntimeState()
        state.port = 3000  # Default port
        
        # Check health endpoint
        try:
            url = f"http://localhost:{state.port}/health"
            req = urllib.request.urlopen(url, timeout=2)
            if req.status == 200:
                state.is_running = True
                state.health_ok = True
        except:
            pass
        
        # Check API endpoints if service is running
        if state.is_running:
            test_endpoints = ["/api/v1", "/api"]
            for endpoint in test_endpoints:
                try:
                    url = f"http://localhost:{state.port}{endpoint}"
                    req = urllib.request.urlopen(url, timeout=2)
                    if req.status in [200, 404]:  # 404 means route exists but no data
                        state.endpoints_responding.append(endpoint)
                except:
                    pass
        
        return state
    
    def _find_discrepancies(self, state: MultiLevelState) -> list[StateDiscrepancy]:
        """Find discrepancies between state levels"""
        discrepancies = []
        
        # Check contract vs code
        for entity in state.contract.entities:
            entity_name = entity.get("name", "")
            if entity_name:
                # Check if model file exists
                model_found = any(
                    entity_name.lower() in f.lower() 
                    for f in state.code.models + state.code.files
                )
                if not model_found:
                    discrepancies.append(StateDiscrepancy(
                        level=StateLevel.CODE,
                        type=DiscrepancyType.MISSING_ENTITY,
                        path=f"models/{entity_name}",
                        expected=entity_name,
                        actual=None,
                        message=f"Entity '{entity_name}' defined in contract but no model file found",
                        severity="warning"
                    ))
        
        # Check required files
        if not state.code.has_package_json:
            discrepancies.append(StateDiscrepancy(
                level=StateLevel.CODE,
                type=DiscrepancyType.MISSING_FILE,
                path="package.json",
                expected="package.json",
                actual=None,
                message="Missing package.json file",
                severity="error"
            ))
        
        # Check runtime vs contract
        if state.contract.entities and not state.runtime.is_running:
            discrepancies.append(StateDiscrepancy(
                level=StateLevel.RUNTIME,
                type=DiscrepancyType.RUNTIME_ERROR,
                path="service",
                expected="running",
                actual="stopped",
                message="Service is not running",
                severity="warning"
            ))
        
        return discrepancies
    
    def write_state_snapshot(self, state: MultiLevelState) -> str:
        """Write state snapshot to file"""
        state_dir = self.target_dir / "state"
        state_dir.mkdir(parents=True, exist_ok=True)
        
        snapshot_path = state_dir / "multi-level-state.json"
        snapshot_path.write_text(json.dumps(state.to_dict(), indent=2))
        
        return str(snapshot_path)
